### ISSUE TITLE
Assertion error when trying to use tf.contrib loss function with keras model

### ISSUE THREAD

**Issue by thebeancounter** [keras-team/keras#11962]
I am using python 3 with anaconda, and trying to use a tf.contrib loss function with a Keras model. 

The code is the following 

    from keras.layers import Dense, Flatten
        from keras.optimizers import Adam
        from keras.models import Sequential
        from tensorflow.contrib.losses import metric_learning
        model = Sequential()
        model.add(Flatten(input_shape=input_shape))
        model.add(Dense(50,  activation="relu"))
        model.compile(loss=metric_learning.triplet_semihard_loss, optimizer=Adam())
        return model

I get the following error: 

> File
> "/home/user/.local/lib/python3.6/site-packages/keras/engine/training_utils.py",
> line 404, in weighted
>     score_array = fn(y_true, y_pred)   File "/home/user/anaconda3/envs/siamese/lib/python3.6/site-packages/tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py",
> line 179, in triplet_semihard_loss
>     assert lshape.shape == 1 AssertionError

When I am using the same network  with a keras loss function it works fine, I tried to wrap the tf loss function in a function like so 

    def func(y_true, y_pred): 
        return tf.contrib.losses.metric_learning.triplet_semihard_loss(y_true, y_pred) 

And still getting the same error 

What am I doing wrong here? 

Also see my SO [question]([url](https://stackoverflow.com/questions/53996020/keras-model-with-tf-contrib-losses-metric-learning-triplet-semihard-loss-asserti))

this is my env 

# Name                    Version                   Build  Channel
_tflow_select             2.3.0                       mkl  
absl-py                   0.6.1                    py36_0  
alabaster                 0.7.12                   py36_0  
anaconda                  custom           py36hbbc8b67_0  
anaconda-client           1.7.2                    py36_0  
anaconda-project          0.8.2                    py36_0  
asn1crypto                0.24.0                   py36_0  
astor                     0.7.1                    py36_0  
astroid                   2.1.0                    py36_0  
astropy                   3.1              py36h7b6447c_0  
atomicwrites              1.2.1                    py36_0  
attrs                     18.2.0           py36h28b3542_0  
babel                     2.6.0                    py36_0  
backcall                  0.1.0                    py36_0  
backports                 1.0                      py36_1  
backports.os              0.1.1                    py36_0  
backports.shutil_get_terminal_size 1.0.0                    py36_2  
beautifulsoup4            4.6.3                    py36_0  
binutils_impl_linux-64    2.31.1               h6176602_1  
binutils_linux-64         2.31.1               h6176602_3  
bitarray                  0.8.3            py36h14c3975_0  
bkcharts                  0.2                      py36_0  
blas                      1.0                         mkl  
blaze                     0.11.3                   py36_0  
bleach                    3.0.2                    py36_0  
blosc                     1.14.4               hdbcaa40_0  
bokeh                     1.0.2                    py36_0  
boto                      2.49.0                   py36_0  
bottleneck                1.2.1            py36h035aef0_1  
bzip2                     1.0.6                h14c3975_5  
c-ares                    1.15.0               h7b6447c_1  
ca-certificates           2018.03.07                    0  
cairo                     1.14.12              h8948797_3  
certifi                   2018.11.29               py36_0  
cffi                      1.11.5           py36he75722e_1  
chardet                   3.0.4                    py36_1  
click                     7.0                      py36_0  
cloudpickle               0.6.1                    py36_0  
clyent                    1.2.2                    py36_1  
colorama                  0.4.1                    py36_0  
conda                     4.5.12                   py36_0  
conda-build               3.17.6                   py36_0  
conda-env                 2.6.0                         1  
contextlib2               0.5.5                    py36_0  
cryptography              2.4.2            py36h1ba5d50_0  
cryptography-vectors      2.4.2                    py36_0  
cudatoolkit               9.0                  h13b8566_0  
cudnn                     7.1.2                 cuda9.0_0  
curl                      7.63.0            hbc83047_1000  
cycler                    0.10.0                   py36_0  
cython                    0.29.2           py36he6710b0_0  
cytoolz                   0.9.0.1          py36h14c3975_1  
dask                      1.0.0                    py36_0  
dask-core                 1.0.0                    py36_0  
datashape                 0.5.4                    py36_1  
dbus                      1.13.2               h714fa37_1  
decorator                 4.3.0                    py36_0  
defusedxml                0.5.0                    py36_1  
distributed               1.25.1                   py36_0  
docutils                  0.14                     py36_0  
entrypoints               0.2.3                    py36_2  
et_xmlfile                1.0.1                    py36_0  
expat                     2.2.6                he6710b0_0  
fastcache                 1.0.2            py36h14c3975_2  
filelock                  3.0.10                   py36_0  
flask                     1.0.2                    py36_1  
flask-cors                3.0.7                    py36_0  
fontconfig                2.13.0               h9420a91_0  
freetype                  2.9.1                h8a8886c_1  
fribidi                   1.0.5                h7b6447c_0  
gast                      0.2.0                    py36_0  
gcc_impl_linux-64         7.3.0                habb00fd_1  
gcc_linux-64              7.3.0                h553295d_3  
get_terminal_size         1.0.0                haa9412d_0  
gevent                    1.3.7            py36h7b6447c_1  
glib                      2.56.2               hd408876_0  
glob2                     0.6                      py36_1  
gmp                       6.1.2                h6c8ec71_1  
gmpy2                     2.0.8            py36h10f8cd9_2  
graphite2                 1.3.12               h23475e2_2  
greenlet                  0.4.15           py36h7b6447c_0  
grpcio                    1.16.1           py36hf8bcb03_1  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb453b48_1  
gxx_impl_linux-64         7.3.0                hdf63c60_1  
gxx_linux-64              7.3.0                h553295d_3  
h5py                      2.9.0            py36h7918eee_0  
harfbuzz                  1.8.8                hffaf4a1_0  
hdf5                      1.10.4               hb1b8bf9_0  
heapdict                  1.0.0                    py36_2  
html5lib                  1.0.1                    py36_0  
icu                       58.2                 h9c2bf20_1  
idna                      2.8                      py36_0  
imageio                   2.4.1                    py36_0  
imagesize                 1.1.0                    py36_0  
importlib_metadata        0.6                      py36_0  
intel-openmp              2019.1                      144  
ipykernel                 5.1.0            py36h39e3cac_0  
ipython                   7.2.0            py36h39e3cac_0  
ipython_genutils          0.2.0                    py36_0  
ipywidgets                7.4.2                    py36_0  
isort                     4.3.4                    py36_0  
itsdangerous              1.1.0                    py36_0  
jbig                      2.1                  hdba287a_0  
jdcal                     1.4                      py36_0  
jedi                      0.13.2                   py36_0  
jeepney                   0.4                      py36_0  
jinja2                    2.10                     py36_0  
jpeg                      9b                   h024ee3a_2  
jsonschema                2.6.0                    py36_0  
jupyter                   1.0.0                    py36_7  
jupyter_client            5.2.4                    py36_0  
jupyter_console           6.0.0                    py36_0  
jupyter_core              4.4.0                    py36_0  
jupyterlab                0.35.3                   py36_0  
jupyterlab_server         0.2.0                    py36_0  
keras                     2.2.4                         0  
keras-applications        1.0.6                    py36_0  
keras-base                2.2.4                    py36_0  
keras-preprocessing       1.0.5                    py36_0  
keyring                   17.0.0                   py36_0  
kiwisolver                1.0.1            py36hf484d3e_0  
krb5                      1.16.1               h173b8e3_7  
lazy-object-proxy         1.3.1            py36h14c3975_2  
libarchive                3.3.3                h5d8350f_5  
libcurl                   7.63.0            h20c2e04_1000  
libedit                   3.1.20170329         h6b74fdf_2  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 8.2.0                hdf63c60_1  
libgfortran-ng            7.3.0                hdf63c60_0  
libgpuarray               0.7.6                h14c3975_0  
libiconv                  1.15                 h63c8f33_5  
liblief                   0.9.0                h7725739_1  
libpng                    1.6.35               hbc83047_0  
libprotobuf               3.6.1                hd408876_0  
libsodium                 1.0.16               h1bed415_0  
libssh2                   1.8.0                h1ba5d50_4  
libstdcxx-ng              8.2.0                hdf63c60_1  
libtiff                   4.0.9                he85c1e1_2  
libtool                   2.4.6                h7b6447c_5  
libuuid                   1.0.3                h1bed415_2  
libxcb                    1.13                 h1bed415_1  
libxml2                   2.9.8                h26e45fe_1  
libxslt                   1.1.32               h1312cb7_0  
llvmlite                  0.26.0           py36hd408876_0  
locket                    0.2.0                    py36_1  
lxml                      4.2.5            py36hefd8a0e_0  
lz4-c                     1.8.1.2              h14c3975_0  
lzo                       2.10                 h49e0be7_2  
mako                      1.0.7                    py36_0  
markdown                  3.0.1                    py36_0  
markupsafe                1.1.0            py36h7b6447c_0  
matplotlib                3.0.2            py36h5429711_0  
mccabe                    0.6.1                    py36_1  
mistune                   0.8.4            py36h7b6447c_0  
mkl                       2018.0.3                      1  
mkl-service               1.1.2            py36h90e4bf4_5  
mkl_fft                   1.0.6            py36h7dd41cf_0  
mkl_random                1.0.1            py36h4414c95_1  
more-itertools            4.3.0                    py36_0  
mpc                       1.1.0                h10f8cd9_1  
mpfr                      4.0.1                hdf1c602_3  
mpmath                    1.1.0                    py36_0  
msgpack-python            0.5.6            py36h6bb024c_1  
multipledispatch          0.6.0                    py36_0  
nbconvert                 5.3.1                    py36_0  
nbformat                  4.4.0                    py36_0  
nccl                      1.3.5                 cuda9.0_0  
ncurses                   6.1                  he6710b0_1  
networkx                  2.2                      py36_1  
ninja                     1.8.2            py36h6bb024c_1  
nltk                      3.4                      py36_1  
nose                      1.3.7                    py36_2  
notebook                  5.7.4                    py36_0  
numba                     0.41.0           py36h962f231_0  
numexpr                   2.6.8            py36hd89afb7_0  
numpy                     1.15.4           py36h1d66e8a_0  
numpy-base                1.15.4           py36h81de0dd_0  
numpydoc                  0.8.0                    py36_0  
odo                       0.5.1                    py36_0  
olefile                   0.46                     py36_0  
openpyxl                  2.5.12                   py36_0  
openssl                   1.1.1a               h7b6447c_0  
packaging                 18.0                     py36_0  
pandas                    0.23.4           py36h04863e7_0  
pandoc                    2.2.3.2                       0  
pandocfilters             1.4.2                    py36_1  
pango                     1.42.4               h049681c_0  
parso                     0.3.1                    py36_0  
partd                     0.3.9                    py36_0  
patchelf                  0.9                  he6710b0_3  
path.py                   11.5.0                   py36_0  
pathlib2                  2.3.3                    py36_0  
patsy                     0.5.1                    py36_0  
pcre                      8.42                 h439df22_0  
pep8                      1.7.1                    py36_0  
pexpect                   4.6.0                    py36_0  
pickleshare               0.7.5                    py36_0  
pillow                    5.3.0            py36h34e0f95_0  
pip                       18.1                     py36_0  
pixman                    0.34.0               hceecf20_3  
pkginfo                   1.4.2                    py36_1  
pluggy                    0.8.0                    py36_0  
ply                       3.11                     py36_0  
prometheus_client         0.5.0                    py36_0  
prompt_toolkit            2.0.7                    py36_0  
protobuf                  3.6.1            py36he6710b0_0  
psutil                    5.4.8            py36h7b6447c_0  
ptyprocess                0.6.0                    py36_0  
py                        1.7.0                    py36_0  
py-lief                   0.9.0            py36h7725739_1  
pycodestyle               2.4.0                    py36_0  
pycosat                   0.6.3            py36h14c3975_0  
pycparser                 2.19                     py36_0  
pycrypto                  2.6.1            py36h14c3975_9  
pycurl                    7.43.0.2         py36h1ba5d50_0  
pyflakes                  2.0.0                    py36_0  
pygments                  2.3.1                    py36_0  
pygpu                     0.7.6            py36h035aef0_0  
pylint                    2.2.2                    py36_0  
pyodbc                    4.0.25           py36he6710b0_0  
pyopenssl                 18.0.0                   py36_0  
pyparsing                 2.3.0                    py36_0  
pyqt                      5.9.2            py36h05f1152_2  
pysocks                   1.6.8                    py36_0  
pytables                  3.4.4            py36h71ec239_0  
pytest                    4.0.2                    py36_0  
pytest-arraydiff          0.3              py36h39e3cac_0  
pytest-astropy            0.5.0                    py36_0  
pytest-doctestplus        0.2.0                    py36_0  
pytest-openfiles          0.3.1                    py36_0  
pytest-remotedata         0.3.1                    py36_0  
python                    3.6.7                h0371630_0  
python-dateutil           2.7.5                    py36_0  
python-libarchive-c       2.8                      py36_6  
pytorch                   0.4.1            py36ha74772b_0  
pytz                      2018.7                   py36_0  
pywavelets                1.0.1            py36hdd07704_0  
pyyaml                    3.13             py36h14c3975_0  
pyzmq                     17.1.2           py36h14c3975_0  
qt                        5.9.7                h5867ecd_1  
qtawesome                 0.5.3                    py36_0  
qtconsole                 4.4.3                    py36_0  
qtpy                      1.5.2                    py36_0  
readline                  7.0                  h7b6447c_5  
requests                  2.21.0                   py36_0  
rope                      0.11.0                   py36_0  
ruamel_yaml               0.15.46          py36h14c3975_0  
scikit-image              0.14.1           py36he6710b0_0  
scikit-learn              0.20.1           py36h4989274_0  
scipy                     1.1.0            py36hfa4b5c9_1  
seaborn                   0.9.0                    py36_0  
secretstorage             3.1.0                    py36_0  
send2trash                1.5.0                    py36_0  
setuptools                40.6.3                   py36_0  
simplegeneric             0.8.1                    py36_2  
singledispatch            3.4.0.3                  py36_0  
sip                       4.19.8           py36hf484d3e_0  
six                       1.12.0                   py36_0  
snappy                    1.1.7                hbae5bb6_3  
snowballstemmer           1.2.1                    py36_0  
sortedcollections         1.0.1                    py36_0  
sortedcontainers          2.1.0                    py36_0  
sphinx                    1.8.2                    py36_0  
sphinxcontrib             1.0                      py36_1  
sphinxcontrib-websupport  1.1.0                    py36_1  
spyder                    3.3.2                    py36_0  
spyder-kernels            0.3.0                    py36_0  
sqlalchemy                1.2.15           py36h7b6447c_0  
sqlite                    3.26.0               h7b6447c_0  
statsmodels               0.9.0            py36h035aef0_0  
sympy                     1.3                      py36_0  
tblib                     1.3.2                    py36_0  
tensorboard               1.12.1           py36he6710b0_0  
tensorflow                1.12.0          mkl_py36h69b6ba0_0  
tensorflow-base           1.12.0          mkl_py36h3c3e929_0  
termcolor                 1.1.0                    py36_1  
terminado                 0.8.1                    py36_1  
testpath                  0.4.2                    py36_0  
theano                    1.0.3            py36hfd86e86_0  
tk                        8.6.8                hbc83047_0  
toolchain                 2.4.0                         0    conda-forge
toolchain_c_linux-64      2.4.0                         0    conda-forge
toolchain_cxx_linux-64    2.4.0                         0    conda-forge
toolz                     0.9.0                    py36_0  
torchvision               0.2.1                    py36_0  
tornado                   5.1.1            py36h7b6447c_0  
tqdm                      4.28.1           py36h28b3542_0  
traitlets                 4.3.2                    py36_0  
typed-ast                 1.1.0            py36h14c3975_0  
unicodecsv                0.14.1                   py36_0  
unixodbc                  2.3.7                h14c3975_0  
urllib3                   1.24.1                   py36_0  
wcwidth                   0.1.7                    py36_0  
webencodings              0.5.1                    py36_1  
werkzeug                  0.14.1                   py36_0  
wheel                     0.32.3                   py36_0  
widgetsnbextension        3.4.2                    py36_0  
wrapt                     1.10.11          py36h14c3975_2  
wurlitzer                 1.0.2                    py36_0  
xlrd                      1.2.0                    py36_0  
xlsxwriter                1.1.2                    py36_0  
xlwt                      1.3.0                    py36_0  
xz                        5.2.4                h14c3975_4  
yaml                      0.1.7                had09818_2  
zeromq                    4.2.5                hf484d3e_1  
zict                      0.1.3                    py36_0  
zlib                      1.2.11               h7b6447c_3  
zstd                      1.3.7                h0b5b093_0 

Thanks

**Comment by gabrieldemarmiesse** [comment]
Tf.contrib is deprecated and will be removed at the next tf release (just so you know) . Maybe you can ask your question to the author of the loss function?

**Comment by thebeancounter** [comment]
So the loss function that i try to use is deprecated? is there a replacement for triplet semi hard loss?

**Comment by gabrieldemarmiesse** [comment]
https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md

Also, issues should be used if you are certain that the mistake is in the keras codebase. Since you didn't write the loss function yourself, it is possible that the mistake is in the loss function, and not in keras itself.

**Comment by msymp** [comment]
Hello @thebeancounter, it looks like you've had a Stack Overflow interaction with Gabriele on this same issue and got your query addressed.
https://stackoverflow.com/questions/53996020/keras-model-with-tf-contrib-losses-metric-learning-triplet-semihard-loss-asserti
Hence we will close this issue. Thanks.
